### ISSUE TITLE
dtype ctor allows "align" kwargs even when unapplicable

### ISSUE THREAD

**Issue by anntzer** [numpy/numpy#8158]
The doc for the `np.dtype` ctor documents its second arg as
```
align : bool, optional
Add padding to the fields to match what a C compiler would output for a similar C-struct.
Can be True only if obj is a dictionary or a comma-separated string. If a struct dtype is
being created, this also sets a sticky alignment flag isalignedstruct.
```
but in fact:
```
In [1]: np.dtype(float, 3)
Out[1]: dtype('float64')
```
i.e. `align` is ignored when not applicable.  Error'ing in such a case would help catching typos, as the user more likely intended to write
```
In [2]: np.dtype((float, 3))
Out[2]: dtype(('<f8', (3,)))
```

**Comment by leofang** [comment]
> Can be True only if obj is a dictionary or a comma-separated string.

@seberg we hit this issue too. The `numpy.dtype()` ctor does not raise any exception when `align=True` is passed with an unsupported format (such as 2-tuples in OP's example), and it's hard to debug why the layout is not as expected. Can we tighten this up?

**Comment by seberg** [comment]
Yeah, sounds fine. I would lean to adding a `DeprecationWarning` and making `align` keyword only (which means all others as well)?
(Rather than trying to tighten the type check.)

Maybe could gamble on libraries not passing by position and going with a `VisibleDeprecationWarning` right away.  (I am not sure a DeprecationWarning would show up in most interesting cases).

**Comment by leofang** [comment]
> Rather than trying to tighten the type check.

What do you mean? I imagine we need to check that a warning (I hope it's `VisibleDeprecationWarning`, given that this bug is always a decade old) is raised for input formats other than a dictionary or a comma-separated string, so somewhere in deciphering the user input we have a logic to resolve the potential overloads, and the warning should be inserted there.

**Comment by seberg** [comment]
I figure this:
```
>>> np.dtype(float, 3)
VisibleDeprecationWarning: `align=3` passed, did you mean to pass a tuple? In the future `align=` must be passed as a keyword.
```
would solve the issue pretty clearly (with grace period).

It seems fine to additionally deprecate `align=3`, plus maybe you are thinking of an additional problem.  Python tends to be forgiving about integers as bools unfortunately, but since `align=nbytes` is plausible, maybe it's best to be strict here and reject integers also (even if `alignment=nbytes` is better if we want to pass a value).

**Comment by seberg** [comment]
@leofang I suspect you mean a different thing (or I am just misunderstanding), but I am not sure what.  `np.dtype(("float", 3))` is always aligned anyway.

Do you have an example of what you were bitten by?

**Comment by seberg** [comment]
Nvm.  A `VisibleDeprecationWarning` for non-bools it is.  I would prefer forcing `align=` because that better rejects the subarray use-case `(dtype, True)` is in a sense a valid subarray dtype.
(And I suppose, I don't see much reason to say `align=True` is wrong to pass for `subarray`.)

But, old pickles include `0` and `1` (fine, those are 10+ year old pickles).  New pickles still incude `False` being passed.
That is wrong, but if we fix that now, maybe it's a bit nicer to users to not break loading old pickles (or warn) until a bit after the change.
